src/util/memory.h

```diff
@@ -16,4 +16,20 @@ std::unique_ptr<T> MakeUnique(Args&&... args)
     return std::unique_ptr<T>(new T(std::forward<Args>(args)...));
 }
 
+/**
+ * Ensure that a global is initialized before its first use and never
+ * destructed, since other destructors might use it.
+ * This class should not be used where the destructor ~T() is non-trivial.
+ */
+template <typename T>
+struct GlobalCOFU {
```

Comment:
This seems over-engineered. Why not just have a simple function like:

```c++
Logger& LogInstance() {
  static Logger logger;
  return logger;
}
```

and have a scripted diff replacing `g_logger->` with `LogInstance().` This would be equivalent to what the `GlobalCOFU` class is doing but more straightforward, and would actually guarantee destruction, and avoid creating landmines where separate variables of the same type point to the same instances, and avoid weird terminology (COFU == singleton?)

Reply:
> and would actually guarantee destruction

FWIW the same could be done with `GlobalCOFU`.